xayn-ai/src/ltr/list_net/mod.rs

```diff
@@ -1037,101 +1135,477 @@ mod tests {
                 evaluation_results: Vec::new(),
             }
         }
     }
 
-    impl Callbacks for TestCallbacks {
-        fn begin_of_batch(&mut self) {
-            dbg!("begin batch");
+    impl TrainingController for TestController {
+        type Error = Infallible;
+
+        type Outcome = (Self, ListNet);
+
+        fn begin_of_batch(&mut self) -> Result<(), Self::Error> {
+            eprintln!("begin batch");
+            Ok(())
         }
 
-        fn end_of_batch(&mut self, losses: Vec<f32>) {
+        fn end_of_batch(&mut self, losses: Vec<f32>) -> Result<(), Self::Error> {
+            eprintln!("end of batch");
             dbg!(losses);
+            Ok(())
         }
 
-        fn begin_of_epoch(&mut self, _list_net: &ListNet) {
-            dbg!("begin epoch");
+        fn begin_of_epoch(
+            &mut self,
+            _nr_batches: usize,
+            _list_net: &ListNet,
+        ) -> Result<(), Self::Error> {
+            eprintln!("begin of epoch");
+            Ok(())
         }
 
         fn end_of_epoch(
             &mut self,
             _list_net: &ListNet,
             mean_kv_divergence_evaluation: Option<f32>,
-        ) {
+        ) -> Result<(), Self::Error> {
+            eprintln!("end of epoch");
             dbg!(mean_kv_divergence_evaluation);
             self.evaluation_results.push(mean_kv_divergence_evaluation);
+            Ok(())
+        }
+
+        fn begin_of_training(
+            &mut self,
+            _nr_epochs: usize,
+            _list_net: &ListNet,
+        ) -> Result<(), Self::Error> {
+            eprintln!("begin of training");
+            Ok(())
         }
 
-        fn begin_of_training(&mut self, _list_net: &ListNet) {
-            dbg!("begin of training");
+        fn end_of_training(&mut self) -> Result<(), Self::Error> {
+            eprintln!("end of training");
+            Ok(())
         }
 
-        fn end_of_training(&mut self, _list_net: &ListNet) {
-            dbg!("end of training");
+        fn training_result(self, list_net: ListNet) -> Result<Self::Outcome, Self::Error> {
+            Ok((self, list_net))
+        }
+    }
+
+    #[test]
+    fn test_training_list_net_is_reproducible_for_same_inputs_and_state() {
+        // drop(crate::embedding::qambert::tests::qambert());
+
+        use Relevance::{High, Low, Medium};
+        let list_net = LIST_NET.clone();
+
+        let inputs = Array1::from(SAMPLE_INPUTS.to_vec())
+            .into_shape((10, 50))
+            .unwrap();
+
+        let relevances = vec![Low, Low, Medium, Medium, Low, Medium, High, Low, High, High];
+        let data_frame = (inputs, prepare_target_prob_dist(&relevances).unwrap());
+
+        let training_data = vec![data_frame.clone(), data_frame.clone(), data_frame.clone()];
+        let test_data = vec![data_frame];
+        let nr_epochs = 5;
+        let batch_size = 1;
+
+        // The NaN bug somehow only appear on the first execution so uncommenting this
```

Comment:
Is this still relevant?

Reply:
I hope not, we didn't really found the bug but happened to work around it.

And while we kinda expect it to be caused by QEMU, it's possible that it is a miscompilation of rust
where rust relies on some kind of implementation detail which happens to work this way on most
hardware but happened to not be implemented that way in QEMU. 

Through I guess I can remove it anyway.